compressor/gzip/gzip.go

```diff
@@ -18,32 +18,197 @@
+// create a gRPC dialer with the Compressor option.
+//
+//  trans := grpc.NewTransport()
+//  dialer := trans.NewDialer(GZIPCompressor)
+//  peers := roundrobin.New(dialer)
+//  outbound := trans.NewOutbound(peers)
+//
+// If you are using the YARPC configurator to create YARPC objects
+// using config files, you will also need to register the compressor
+// with your configurator.
+//
+//  configurator := yarpcconfig.New()
+//  configurator.MustRegisterCompressor(GZIPCompressor)
+//
+// Then, using the compression strategy for outbound requests
+// on a particular client, just set the compressor to gzip.
+//
+//  outbounds:
+//    theirsecureservice:
+//      grpc:
+//        address: ":443"
+//        tls:
+//          enabled: true
+//        compressor: gzip
+//
+func New(opts ...Option) *Compressor {
+	c := &Compressor{
+		level: gzip.DefaultCompression,
+	}
+	for _, opt := range opts {
+		opt.apply(c)
+	}
+	return c
+}
+
 // Compressor represents the gzip compression strategy.
-type Compressor struct{}
+type Compressor struct {
+	level         int
+	compressors   sync.Pool
+	decompressors sync.Pool
+}
 
-var _ transport.Compressor = Compressor{}
+var _ transport.Compressor = (*Compressor)(nil)
 
 // Name is gzip.
-func (Compressor) Name() string {
-	return "gzip"
+func (*Compressor) Name() string {
+	return name
 }
 
 // Compress creates a gzip compressor.
-func (Compressor) Compress(w io.Writer) (io.WriteCloser, error) {
-	return gzip.NewWriter(w), nil
+func (c *Compressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	cw, got := c.compressors.Get().(*writer)
+	if !got {
+		cw, err := gzip.NewWriterLevel(w, c.level)
+		if err != nil {
+			return nil, err
+		}
+		return &writer{
+			writer: cw,
+			pool:   &c.compressors,
+		}, nil
+	}
+	cw.writer.Reset(w)
+	return cw, nil
+}
+
+type writer struct {
+	writer *gzip.Writer
+	pool   *sync.Pool
 }
 
-// Decompress creates a gzip decompressor.
-func (Compressor) Decompress(r io.Reader) (io.Reader, error) {
-	return gzip.NewReader(r)
+var _ io.WriteCloser = (*writer)(nil)
+
+func (w *writer) Write(buf []byte) (int, error) {
+	return w.writer.Write(buf)
+}
+
+func (w *writer) Close() error {
+	defer w.pool.Put(w)
+	return w.writer.Close()
+}
+
+// Decompress obtains a gzip decompressor.
+func (c *Compressor) Decompress(r io.Reader) (io.Reader, error) {
+	dr, got := c.decompressors.Get().(*reader)
+	if !got {
+		dr, err := gzip.NewReader(r)
+		if err != nil {
+			return nil, err
+		}
+		return &reader{
+			reader: dr,
+			pool:   &c.decompressors,
+		}, nil
+	}
+
+	if err := dr.reader.Reset(r); err != nil {
+		// Discard pooled decompressor if it fails to reset to avoid polluting
+		// the pool.
+		return nil, err
+	}
+
+	return dr, nil
+}
+
+type reader struct {
+	reader *gzip.Reader
+	pool   *sync.Pool
+}
+
+var _ io.Reader = (*reader)(nil)
+
+func (r *reader) Read(buf []byte) (n int, err error) {
+	n, err = r.reader.Read(buf)
+	if err == io.EOF {
+		r.pool.Put(r)
```

Comment:
I could allow for premature close. I’d need to write a method to attenuate a Compressor to a GRPCCompressor so it could be registered. gRPC would not use it, but we’d have more liberty in a future version.

Reply:
Yeah, it might be a cleaner approach longer term to add a `Close` method, and only put the reader back in the pool on `Close()`. If we do that, in gRPC, we'd need the wrapper to do the `if err == io.EOF { Close(..) }` though

Reply:
Nice to see that I was not the only one who was confused about gRPC not using the Close method. SGTM.